add `keep_quoted=strict` mode

syranide opened the pull request:
Preliminary PR for #3153 

Adds a new mode for `keep_quoted` which mangles non-quoted property names regardless of them being quoted somewhere else, i.e. it does not reserve quoted property names. This is helpful for further reducing file size and also ensuring more complete obfuscation of code.

PS. Obviously tests needs to be added and updated, etc. I also did this PR against the harmony branch because that's what I rely on, but I can obviously make it for master instead, or both.

**EDIT:** PR redone by updating the AST_Dot-definition instead.

kzc commented:
The removal of the `"properties"` option code from `OPT(AST_Sub)` caused a number of tests to fail. Existing functionality shouldn't break. `minify` and `CLI` tests covering typical scenarios would also be needed.

I'm not convinced performing transforms in the output module is the right approach.

Before pursuing further work on this PR you'd probably want to get @alexlamsl's input to see if it's something that would be considered. I'll step out of this discussion and let you two hash it out.

syranide commented:
@kzc Updated the PR. I tried it by simply updating the definition for `AST_Dot` instead as you suggested, adding a `quote` prop to it. Very very simple and breaks nothing out-of-the-box. Maybe it makes more sense to do it on `AST_PropAccess` as you said, but then `quote` becomes duplicated in `AST_Sub), anyway you guys are the experts on that, I'll gladly update it to whatever you suggest. What can I say though is that this solution works very well while being minimally invasive.

I'll wait for @alexlamsl to leave his feedback.

(PS. I'll add tests if you greenlight this PR).

alexlamsl commented:
The only functional changes that are acceptable for this feature (no opinion on its usefulness or usability) would be to `lib/propmangle.js` only.

I'm expecting to review `lib/propmangle.js`, `test/compress/properties.js` and `README.md`

kzc commented:
Should rebase this PR off of `master` since `uglify-es` (harmony) is no longer maintained. 

If this PR is accepted for `uglify-js` it may be considered for [terser](https://github.com/fabiosantoscode/terser).

kzc commented on lib/propmangle.js:
```diff
@@ -158,31 +158,45 @@ function mangle_properties(ast, options) {
     ast.walk(new TreeWalker(function(node){
         if (node instanceof AST_ObjectKeyVal) {
-            add(node.key);
+            if (options.keep_quoted !== "strict" || !node.quote) {
```
This condition and the others can be made more efficient by declaring the following above the step 1 `ast.walk`:
```js
    var keep_quoted_strict = options.keep_quoted === "strict";
```

syranide commented:
@alexlamsl As far as I can make out, it is impossible to do it strictly within those parameters without resorting to what I would intuitively consider hacks. But I would say on the surface the current implementation is in the spirit of what you're asking. I'll explain the non-conforming changes:

`propmangle.reserve_quoted_keys` needs to stop being run but it does not have access to mangle-options. So the simplest solution is to just add it to the if-statement in `minify.js`, which seems fair given how `options.mangle.properties.keep_quoted` is referenced elsewhere in the file.

The other problem is that after `compress` has taken place, information regarding the original source of `AST_Dot` and whether it was a quoted `AST_Sub` is irretrievably lost. So AFAIK there's four options;

1. Annotate the AST. Straight-forward, what this PR implements.
2. Move `AST_Sub=>AST_Dot` transform from `compress` to `output`. Rather invasive and complex.
3. Running mangling before compression. Suboptimal results as @kzc pointed out.
4. Perform mangling in two steps, pre (mark nodes for mangling) and post (mangle nodes). But this is even more invasive.

This PR implements nr1 which involves adding a `quote` prop to `AST_Dot` and also setting `quote` in the related transform in `compress` so that the information is available when `mangleprops` is run. Perhaps another node-type should be annotated, I'm not the expert on that.

PS. Didn't know harmony was deprecated, changed base to master. So I've updated the the PR with all the above in-mind and it passes all existing tests (I will add related tests too). I don't mind another solution for the AST-annotation, but I require your preference in that matter then.

kzc commented on lib/propmangle.js:
```diff
@@ -135,38 +135,51 @@ function mangle_properties(ast, options) {
 
-    // step 1: find candidates to mangle
     ast.walk(new TreeWalker(function(node){
+        var keep_quoted_strict = options.keep_quoted === "strict";
```
`var keep_quoted_strict` should be moved before the step 1`walk` so it doesn't incur a cost for each node visited. That way the declaration need not be repeated later at line 166.

Also, please don't remove the comment.

kzc commented on lib/propmangle.js:
```diff
@@ -135,38 +135,51 @@ function mangle_properties(ast, options) {
-        else if (node instanceof AST_Call
-            && node.expression.print_to_string() == "Object.defineProperty") {
-            addStrings(node.args[1], add);
+            if (!keep_quoted_strict) {
```
Please do not remove existing functionality. The code is there for a reason.

kzc commented on lib/compress.js:
```diff
@@ -6208,7 +6208,8 @@ merge(Compressor.prototype, {
                         expression: expr,
-                        property: property
+                        property: property,
+                        quote: prop.quote
```
Trailing commas are acceptable in new code in order to minimize future diffs.

syranide commented:
@alexlamsl No stress, but feel free to review this PR whenever you have the time. https://github.com/mishoo/UglifyJS2/pull/3158#issuecomment-393169316 explains the rationale but I believe this is as good as I can get it without very specific directions/opinions. So I guess it's a question of whether the required changes are acceptable or not.

kzc commented:
Tests are needed.

syranide commented:
@kzc Considering he seems on the fence on the scope of it I'd rather not waste time writing tests if it's not going to be merged in the end. If the approach is acceptable then I don't mind writing a few tests.

kzc commented:
Having tests would increase the likelihood of review. In any case this PR is simpler and less invasive than the previous approach.

syranide commented:
I've added a basic test which should cover all *apparent* cases.

syranide commented:
Fixed.

kzc commented:
Please use an actual value for `expect_stdout` rather than `true` and write the test in a standalone manner without `window` properties. Use `console.log` in the test in such a way that it outputs something meaningful to stdout dependent on the new functionality being tested.  See other tests for examples. 

Try to use double quotes rather than single quotes and format the expected result with `uglifyjs -b`.

syranide commented:
> Try to use double quotes rather than single quotes and format the expected result with uglifyjs -b.

Done and sure, but I'm not sure how to get it to disable all default transforms (to be compatible with the options used in the test), there doesn't seem to be a flag for it?

kzc commented on test/compress/keep_quoted_strict.js:
```diff
@@ -0,0 +1,39 @@
+    expect: {
+        var a={p:1,get r(){return 2},propc:3,get propd(){return 4}};
+        var b={propa:5,get propb(){return 6},propc:7,get propd(){return 8}};
+        console.log(a.p,a.r,a.propc,a.propc,a.propd,a.propd);
```
please reformat expected result with `uglifyjs -b`

kzc commented on test/run-tests.js:
```diff
@@ -114,7 +114,9 @@ function run_compress_tests() {
                 if (!Array.isArray(quoted_props)) quoted_props = [];
                 test.mangle.properties.reserved = quoted_props;
-                U.reserve_quoted_keys(input, quoted_props);
+                if (test.mangle.properties.keep_quoted !== 'strict') {
```
please use double quotes

kzc commented on test/compress/keep_quoted_strict.js:
```diff
@@ -0,0 +1,39 @@
+        Object.defineProperty({},"p",{});
+        Object.defineProperty({},"propc",{});
+    }
+    expect_stdout: "1 2 3 3 4 4\n5 6 7 7 8 8"
```
Please use array form for multiline results - see other examples.

kzc commented:
An exception to the double quote coding convention is when a test requires the testing of single quotes. See `-b quote_style=3` to preserve original quote style in output.

kzc commented on test/compress/keep_quoted_strict.js:
```diff
@@ -0,0 +1,39 @@
+        console.log(a.propa, a.propb, a.propc, a["propc"], a.propd, a['propd']);
+        console.log(b["propa"], b["propb"], b.propc, b["propc"], b.propd, b['propd']);
+        Object.defineProperty({}, "propa", {});
+        Object.defineProperty({}, "propc", {});
```
Please incorporate these `Object.defineProperty` calls into the program flow with a consequential impact on `console.log` output. Move these two lines above the first `console.log` and have one statement operate on object `a` and the other on object `b`.

syranide commented:
@kzc Pushed changes, I'm not sure if I really see the point in the logging, the test should only be concerned with the code transformation being correct. The behavior will not change by definition.

Anyway, it's pushed. However, as I mentioned above `uglify -b` is not useful, it enables all default transformations which emits an entirely different output. Can I get rid of the default transforms somehow?

kzc commented:
The reason for the test console logging is that each test with `expect_stdout` is re-run with a variety of minify options from `test/ufuzz.json`. This amplifies the code coverage of each test case and has been very useful in finding bugs.

[`terser`](https://www.npmjs.com/package/terser), a maintained fork of `uglify-es`, has a `defaults` options for `compress` that you could use for formatting this test case...

```
$ echo 'if(1)console.log(1+2);' | terser -b -c defaults=false,conditionals=true
1 && console.log(1 + 2);

$ echo 'if(1)console.log(1+2);' | terser -b -c defaults=false,evaluate=true
if (1) console.log(3);
```

syranide commented:
@kzc Well that doesn't really work either as I need to include the new `keep_quote=strict` transformation too...

syranide commented:
@kzc Just made the changes by hand instead. I see your point regarding fuzzing :+1:

kzc commented on README.md:
```diff
@@ -807,6 +807,9 @@ UglifyJS.minify(code, { mangle: { toplevel: true } }).code;
 - `keep_quoted` (default: `false`) -— Only mangle unquoted property names.
+  - `true` -- quoted property names are reserved and any unquoted properties with the
+    same property name will not be mangled.
+  - `strict` -- advanced, all unquoted property names are mangled unless reserved.
```
Please double quote `"strict"` here, as it's a string.

syranide commented:
Ping

knaccc commented:
Thanks for implementing this, it's an essential feature for me. The reason is that without this feature, things break badly in the following common scenario:

a.js:
```
var lib = {
  hello : function() {
    console.log('hello');
  }
};
```
which mangles to:
```
var l={l:function(){console.log("hello")}};
```
b.js:

```
var a = {};
a['hello'] = 123;
lib.hello();
```
which mangles to:
```
var a={};a["hello"]=123;l.hello();
```

This means that even when `--name-cache` is used, this results in the mangled b.js attempting to call `l.hello()` instead of `l.l()`, which means the code breaks.

Could anyone provide an update on when this feature might become merged into the codebase please?

syranide commented:
I'm sorry if this is frowned upon, @knaccc it's been accepted and released in Terser (ES6 uglify-fork). Although I'm not using nameCache myself and haven't tested it for that, I would assume it works. It's an interesting obersevation that nameCache becomes file-order dependent (which is really bad) when used with the regular property mangling behavior.

knaccc commented:
@syranide Thank you so much for telling me about Terser - that looks like a perfect solution for me, especially because I wanted to use ES6 features and thought I'd have to abandon that idea because of a lack of any decent mangler.

I've just tested again with `--name-cache`, and it doesn't work, no matter which way around it's done.

For the example I gave, this is what name-cache.json looks like after b.js is mangled:

`{"vars":{"props":{"$a":"a"}},"props":{"props":{}}}`

And then after a.js is mangled, it looks like this:

`{"vars":{"props":{"$a":"a","$lib":"l"}},"props":{"props":{"$hello":"l"}}}`

syranide commented:
@knaccc Ah, the name-cache does not keep track of the list of reserved identifiers added by the standard property mangling behavior. Again, I haven't used it myself, but at first glance it seems as if name-cache is indeed broken when used with the standard property mangling. Even if that was fixed it would still be order dependent and the only solution I see would be a pre-pass like you tried above.
